addon/addon/models/animation-participant.ts

```diff
@@ -32,39 +32,128 @@ export class AnimationParticipantManager {
   // In which case we don't want to call it during updateParticipants but instead
   // have the animations service call this
   performCleanup() {
-    for (let animationParticipant of this.participants) {
-      // Naive cleanup for now.
-      // This should be changed to something that can prune DOMRefs AND graft live nodes
-      // It should collect all DOMRefs to dispose, then traverse the tree and remove them at one go, preserving
-      // nodes that are NOT disposed and all of their descendants
-      animationParticipant._DOMRefsToDispose.forEach((DOMRef) => {
-        DOMRef.delete();
-      });
+    let unanimatedDetachedDOMRefs = new Set<DOMRefNode>();
+    let displacedDOMRefs = new Set<DOMRefNode>();
+    for (let participant of this.participants) {
+      participant._DOMRefsToDispose.forEach((node) =>
+        displacedDOMRefs.add(node)
+      );
+      if (
+        participant.uiState.detached &&
+        (!participant.uiState.detached.animation ||
+          participant.uiState.detached.animation.playState === 'finished' ||
+          participant.uiState.detached.animation?.playState === 'idle' ||
+          participant.uiState.detached.animation?.playState === 'paused')
+      ) {
+        // TODO: when cloning is implemented, add the DOMRef here
+        unanimatedDetachedDOMRefs.add(participant.uiState.detached.DOMRef);
+      }
+    }
 
+    let deleted: DOMRefNode[] = [];
+    let nodesToGraft = new Map<DOMRefNode, DOMRefNode>();
+    function recurse<T>(
+      node: DOMRefNode,
+      callback: (node: DOMRefNode, scope: T) => { nextScope: T; stop: boolean },
+      scope: T
+    ) {
+      let { nextScope, stop } = callback(node, scope);
+      if (stop) return;
+      for (let child of node.children) {
+        recurse(child, callback, nextScope);
+      }
+    }
+    for (let node of displacedDOMRefs) {
+      recurse<undefined>(
+        node,
+        (node, _) => {
+          if (node.parent) node.delete();
+          else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+          deleted.push(node);
+          return {
+            nextScope: undefined,
+            stop: false,
+          };
+        },
+        undefined
+      );
+    }
```

Comment:
Displaced DOMRefs are DOMRefs that were detached, and then another detached DOMRef came in. They should not be in use and we should pluck them all out of the tree, including their descendants. Interesting thing is their descendants may not get displaced if there is no equivalent descendant in the element that displaced them, so we have to remove the descendants indiscriminately.